rfcs/0108-nixos-containers.md

```diff
@@ -0,0 +1,467 @@
+[`systemd-nspawn@.service`](https://github.com/systemd/systemd/blob/v247/units/systemd-nspawn@.service.in) will be used.
+
+The approach how a container is bootstrapped won't change and will thus consist of the
+following steps (executed via a custom `ExecStartPre=`-script):
+
+* Create an empty directory in `/var/lib/machines` named like the container-name.
+* `systemd-nspawn` only expects `/etc/os-release`, `/etc/machine-id` and `/var` to exist
+  inside, however with no content.
+* To get a running NixOS inside, `/nix/store` is bind-mounted into it. As an `init` process,
+  the [stage-2 script](https://github.com/NixOS/nixpkgs/blob/master/nixos/modules/system/boot/stage-2-init.sh)
+  is started which eventually `exec(2)`s into `systemd` and ensures that everything
+  is correctly set up.
+* The option `boot.isContainer = true;` will be automatically set for new containers as well.
+  This is necessary to
+  * avoid bogus `modprobe` calls since `nspawn` doesn't have its own kernel.
+  * as option to skip building a `stage-1` boot script when building a NixOS system for
+    the container.
+
+This init-script can be built by evaluating a NixOS config against `<nixpkgs/nixos/lib/eval-config.nix>`.
+
+Support for existing tarballs to be imported with `machinectl pull-tar` is explicitly out of
+scope in this RFC.
+
+## Network
+
+The following section provides an overview of how to configure networking for containers and
+how this will be implemented. A proposal how the API of the NixOS module could look like will
+be demonstrated in the [next chapter](#examples-and-interactions).
+
+### "public" networking
+
+This is the most trivial networking mode. It is taken if the `PrivateNetwork`-option of the
+`.nspawn`-unit is set to `no`. In this case, the container has full access to the host's network,
+otherwise the container will run in its own namespace.
+
+### Default Mode
+
+If nothing else is specified, the [default settings of `systemd-nspawn`](https://github.com/systemd/systemd/blob/v247/network/80-container-ve.network) will
+be used for networking. To briefly summarize, this means:
+
+* A [`veth`](https://man7.org/linux/man-pages/man4/veth.4.html) interface-pair will be created,
+  one "host-side" interface and a container interface inside its own namespace.
+* A dynamically allocated prefix of `0.0.0.0/28` (or `0.0.0.0/24` for virtual zones) will be used
+  as address pool to distribute IPv4 addresses via DHCP to containers.
+
+Additionally, basic IPv6 support was implemented:
+
+* By specifying `::/64`, a [RFC4193 ULA prefix](https://tools.ietf.org/html/rfc4193) will be
+  allocated to the host-side interface.
+* With [`radvd`](https://github.com/radvd-project/radvd), containers can assign themselves addresses from this address prefix by utilizing
+  [RFC4862 SLAAC](https://tools.ietf.org/html/rfc4862)
+
+This is necessary since `systemd-networkd` doesn't support router advertisements with
+dynamically allocated prefixes.
```

Comment:
Same same, I'm not sure if IPv6 prefix delegation on this is a thing/will be in the future. It shouldn't be part of an RFC.